source/Halibut.Tests/BiDirectionalTests.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Halibut.ServiceModel;
+using Halibut.Tests.Support;
+using Halibut.Tests.TestServices;
+using Halibut.Tests.TestServices.Async;
+using Halibut.TestUtils.Contracts;
+using Microsoft.VisualStudio.Threading;
+using NUnit.Framework;
+
+namespace Halibut.Tests
+{
+    public class BiDirectionalTests : BaseTest
+    {
+        /// <summary>
+        /// Test is redundant but kept around since we really want the security part to work. 
+        /// </summary>
+        [Test]
+        public async Task ClientCanRespondToServerRequests()
+        {
+            await SetupBiDirectionalClients(async (octopusEchoClient, tentacleEchoClient) =>
+            {
+                var octopusResponse = await octopusEchoClient.SayHelloAsync("Hello");
+                var tentacleResponse = await tentacleEchoClient.SayHelloAsync("World");
+
+                octopusResponse.Should().Be("Hello...");
+                tentacleResponse.Should().Be("World...");
+            });
+        }
+
+        [Test]
+        public async Task ConcurrentTasksDoNotBlock()
+        {
+            await SetupBiDirectionalClients(async (octopusEchoClient, tentacleEchoClient) =>
+            {
+                var sw = Stopwatch.StartNew();
+                var t1 = Task.Run(async () => await RunLongRunningTask(octopusEchoClient)).WithTimeout(TimeSpan.FromSeconds(12));
+                var t2 = Task.Run(async () => await RunLongRunningTask(tentacleEchoClient)).WithTimeout(TimeSpan.FromSeconds(12));
+
+                await Task.WhenAll(t1, t2);
+                sw.Stop();
+
+                sw.ElapsedMilliseconds.Should().BeLessThan(await t1 + await t2);
+            });
+            
+            async Task<long> RunLongRunningTask(IAsyncClientEchoService tentacleEchoClient)
+            {
+                var longRunningDuration = Stopwatch.StartNew();
+                await tentacleEchoClient.LongRunningOperationAsync();
+                longRunningDuration.Stop();
+                return longRunningDuration.ElapsedMilliseconds;
+            }
+        }
+
+        [Test]
+        public async Task ConcurrentClientAndServerRequestsCorrectlyInterleaved()
+        {
+            await SetupBiDirectionalClients(async (octopusEchoClient, tentacleEchoClient) =>
+            {
+                var t1 = Task.Run(async () => await RunEchoTask(octopusEchoClient)).WithTimeout(TimeSpan.FromSeconds(5));
```

Comment:
What is the intent of the timeout? The base class should be providing a timeout on the test if the concern is this will run forever.

Reply:
👍 I've update the timeout limits for use in the test,